src/backend/joanie/payment/models.py

```diff
@@ -1 +1,465 @@
+                    & models.Q(total__gte=0)
+                )
+                | (models.Q(order__isnull=True) & models.Q(parent__isnull=False)),
+                name="invoice_relies_on_order_or_invoice",
+            ),
+        ]
+
+    def __str__(self):
+        types = dict(payment_enums.INVOICE_TYPES)
+        return f"{types[self.type]} {self.reference}"
+
+    @property
+    def state(self):
+        """
+        Process the state of the invoice
+        """
+        if self.balance.amount >= 0:
+            if (
+                self.invoiced_balance.amount == 0
+                and self.transactions_balance.amount == 0
+            ):
+                return payment_enums.INVOICE_STATE_REFUNDED
+
+            return payment_enums.INVOICE_STATE_PAID
+
+        return payment_enums.INVOICE_STATE_UNPAID
+
+    @property
+    def related_order(self):
+        """
+        Return the related order.
+        """
+        return self.order or self.parent.order
+
+    @property
+    def document(self):
+        """
+        Get the document related to the invoice instance;
+        """
+        # pylint: disable=no-member
+        document = InvoiceDocument(context_query=self.get_document_context())
+        return document.create(persist=False)
+
+    @property
+    def type(self):
+        """
+        Return the invoice type according to its total amount.
+        If total amount is positive, invoice type is "invoice"
+        otherwise it's "credit_note"
+        """
+        if self.total.amount < 0:  # pylint: disable=no-member
+            return payment_enums.INVOICE_TYPE_CREDIT_NOTE
+
+        return payment_enums.INVOICE_TYPE_INVOICE
+
+    @cached_property
+    def transactions_balance(self):
+        """
+        Process the transactions balance.
+
+        First we retrieve all transactions registered for the current invoice
+        and its children. Then we sum all transactions amount.
+        """
+        invoices = [self, *self.children.all()]
+        ids = [invoice.id for invoice in invoices]
+
+        amount = Transaction.objects.filter(invoice__id__in=ids).aggregate(
+            total=models.Sum("total")
+        )["total"] or D(0.00)
+
+        return Money(amount, self.total.currency)  # pylint: disable=no-member
+
+    @cached_property
+    def invoiced_balance(self):
+        """
+        Process the invoiced amount.
+
+        First we retrieve all invoice's children then we sum all invoice amount.
+        """
+        invoices = [self, *self.children.all()]
+        amount = sum(invoice.total.amount for invoice in invoices)
+
+        return Money(amount, self.total.currency)  # pylint: disable=no-member
+
+    @cached_property
+    def balance(self):
+        """
+        Difference between transaction balance and invoiced balance.
+        """
+        # - Compute the balance
+        return self.transactions_balance - self.invoiced_balance
+
+    def _update_context(self):
```

Comment:
update_context makes me think that there is an existing context that we modify. In fact it seems that you are creating the context?

Reply:
after reading what follows maybe this should be called `set_localized_context`?